File: shortest.py
class WordDistance:
    def __init__(self, wordsDict: List[str]):
        self.wordsDict = wordsDict
        self.locations = defaultdict(list)
        # Prepare a mapping from a word to all it's locations (indices).
        for i, w in enumerate(wordsDict):
            self.locations[w].append(i)
        # print(self.locations)

    def shortest(self, word1: str, word2: str) -> int:
        loc1, loc2 = self.locations[word1], self.locations[word2]
        l1, l2 = 0, 0
        min_diff = float("inf")

        # Until the shorter of the two lists is processed
        while l1 < len(loc1) and l2 < len(loc2):
            min_diff = min(min_diff, abs(loc1[l1] - loc2[l2]))
            if loc1[l1] < loc2[l2]:
                l1 += 1
            else:
                l2 += 1
        return min_diff
    
    def bruteForceshortest(self, word1: str, word2: str) -> int:
        l, r = -1, -1
        dist = len(self.wordsDict)
        # print("self.wordsDict", self.wordsDict  )
        for idx, word in enumerate(self.wordsDict):
            if word == word1:
                l = idx
            elif word == word2:
                r = idx
            if l != -1 and r != -1:
                dist = min(dist , abs(l-r))
        return dist
        


# Your WordDistance object will be instantiated and called as such:
# param_1 = obj.shortest(word1,word2)

File: shortest.js
// Time Complexity : O(m+k) where m and k is no. of times word1 and word2 occuring in wordsDict array
// Space Complexity : O(n)

let hashMap; 

/**
 * @param {string[]} wordsDict
 */
var WordDistance = function(wordsDict) {
    hashMap = new Map();
    for(let i = 0; i < wordsDict.length; i++) {
        if(!hashMap.has(wordsDict[i])) {
            hashMap.set(wordsDict[i],[]);
        }
        hashMap.get(wordsDict[i]).push(i);
    }
};

/** 
 * @param {string} word1 
 * @param {string} word2
 * @return {number}
 */
WordDistance.prototype.shortest = function(word1, word2) {
    let list1 = hashMap.get(word1);
    let list2 = hashMap.get(word2);
    let p1=0, p2=0;
    let minDist = Infinity;
    while(p1 !== list1.length && p2 !== list2.length) {
        minDist = Math.min(minDist, Math.abs(list2[p2] - list1[p1]));
        if(minDist === 1) break;
        if(list1[p1] < list2[p2]) p1++;
        else p2++;
    }
    return minDist;
};

/** 
 * Your WordDistance object will be instantiated and called as such:
 * var obj = new WordDistance(wordsDict)
 * var param_1 = obj.shortest(word1,word2)
 */